src/formats/MMTF.cpp

```diff
@@ -238,172 +248,259 @@ void MMTFFormat::read(Frame& frame) {
             // This is a string in MMTF, differs from the name as then increments linearly
             // For example, the fourth chainid in ( A B A B) would be D, not B (the chainname)
             residue.set("chainid", structure_.chainIdList[chainIndex_]);
-
-            // An integer
             residue.set("chainindex", chainIndex_);
-
-            if (!structure_.chainNameList.empty()) {
+            if (!mmtf::isDefaultValue(structure_.chainNameList)) {
                 residue.set("chainname", structure_.chainNameList[chainIndex_]);
             }
 
             frame.add_residue(std::move(residue));
             groupIndex_++;
         }
-
         chainIndex_++;
     }
-
     modelIndex_++;
+
+    // Add additional global (not by group) bonds
     for (size_t i = 0; i < structure_.bondAtomList.size() / 2; i++) {
         auto atom1 = static_cast<size_t>(structure_.bondAtomList[i * 2]);
         auto atom2 = static_cast<size_t>(structure_.bondAtomList[i * 2 + 1]);
 
         // We are below the atoms we care about
         if (atom1 < atomSkip_ || atom2 < atomSkip_) {
+            if (!(atom1 < atomSkip_ && atom2 < atomSkip_)) {
+                warning("MMTF Reader", "chemfiles can not represent bonds between different models");
+            }
             continue;
         }
 
         // We are above the atoms we care about
         if (atom1 > atomIndex_ || atom2 > atomIndex_) {
+            if (!(atom1 > atomIndex_ && atom2 > atomIndex_)) {
+                warning("MMTF Reader", "chemfiles can not represent bonds between different models");
+            }
             continue;
         }
 
-        size_t atom_idx1 = atom1 - atomSkip_;
-        size_t atom_idx2 = atom2 - atomSkip_;
-
-        frame.add_bond(atom_idx1, atom_idx2);
+        frame.add_bond(atom_id(atom1), atom_id(atom2));
     }
 
     atomSkip_ = atomIndex_;
 }
 
 void MMTFFormat::write(const Frame& frame) {
-    // Used to add bonds  bonds
-    auto writeAtomLoc = static_cast<size_t>(structure_.numAtoms);
-
     structure_.numModels++;
     structure_.chainsPerModel.emplace_back(0);
+    structure_.numAtoms += static_cast<int32_t>(frame.size());
```

Comment:
Nitpick: Should a check for exceding the possible size of an MMTF file be placed here? I know that it is unlikely to happen, but this is done for the **PDB** and **GRO** formats already.

Reply:
You mean exceeding `int32_t`? Or is there another size limit for MMTF?

Reply:
Exceding `int32_t`. Probably not needed.